firestore.rules

```diff
@@ -13,8 +13,17 @@ service cloud.firestore {
       allow read: if request.auth.uid != null && request.auth.token.firebase.sign_in_provider != 'anonymous';
     }
     match /sources/{source}/answers/{document=**} {
-      allow read: if request.auth.token.user_type == 'teacher' && request.auth.token.class_hash == resource.data.context_id ||
-                     request.auth.token.user_type == 'learner' && string(request.auth.token.platform_user_id) == string(resource.data.platform_user_id);
+      allow read: if request.auth.token.user_type == 'teacher' && request.auth.token.class_hash == resource.data.context_id;
+      allow read: if request.auth.token.user_type == 'learner' && string(request.auth.token.platform_user_id) == string(resource.data.platform_user_id);
+      allow update: if request.auth.token.user_type == 'learner'
+              && string(request.auth.token.platform_user_id) == string(resource.data.platform_user_id)
+              && request.auth.token.class_hash == resource.data.context_id;
+      allow create: if request.auth.token.user_type == 'learner'
+              && string(request.auth.token.platform_user_id) == string(request.resource.data.platform_user_id)
+              && request.auth.token.class_hash == request.resource.data.context_id;
+      // anonymous requests based on run_key, which must be larger than 10 chars
+      allow read, update: if resource.data.run_key.size() > 10;
```

Comment:
Don't we also want to match the `run_key` in this case, e.g. something like `request.resource.data.run_key == resource.data.run_key`?

Reply:
I'd second that, it seems this would be useful to have this check here, so one anonymous user can't update somebody else's answer.

Reply:
D'oh, of course you're right.